service/history/hsm/operation_logger.go

```diff
@@ -0,0 +1,349 @@
+			tag.Name("oplog"),
+			tag.Value("capacity_reached"))
+		l.operations = l.operations[1:]
+	}
+
+	l.operations = append(l.operations, opCopy)
+	return nil
+}
+
+// TODO: Consider memory allocation optimizations:
+// - Consider returning an iterator instead
+func (l *OpLog) GetOperations(ctx context.Context, path []string) ([]*persistencespb.StateMachineOperation, error) {
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	default:
+	}
+
+	l.RLock()
+	defer l.RUnlock()
+
+	if l.closed {
+		return nil, fmt.Errorf("operation log is closed")
+	}
+
+	l.logger.Debug("Fetching operations",
+		tag.Name("oplog"),
+		tag.Value(fmt.Sprintf("path: %v", path)))
+
+	// Pre-allocate with reasonable size
+	matches := make([]*persistencespb.StateMachineOperation, 0, min(len(l.operations), maxResultsLimit))
+
+	// Break up work into chunks and check context periodically
+	const chunkSize = 1000
+	for i := 0; i < len(l.operations) && len(matches) < maxResultsLimit; i += chunkSize {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		end := i + chunkSize
+		if end > len(l.operations) {
+			end = len(l.operations)
+		}
+
+		for _, op := range l.operations[i:end] {
+			if pathEquals(op.Path, path) {
+				matches = append(matches, proto.Clone(op).(*persistencespb.StateMachineOperation))
+				if len(matches) >= maxResultsLimit {
+					l.logger.Warn("GetOperations reached result limit",
+						tag.Name("oplog"),
+						tag.Value(fmt.Sprintf("limit: %d", maxResultsLimit)))
+					break
+				}
+			}
+		}
+	}
+
+	return matches, nil
+}
+
+func (l *OpLog) PruneOperations(ctx context.Context, retention time.Duration) error {
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
+
+	l.Lock()
+	defer l.Unlock()
+
+	if l.closed {
+		return fmt.Errorf("operation log is closed")
+	}
+
+	beforeCount := len(l.operations)
+
+	cutoff := time.Now().Add(-retention)
+	var retained []*persistencespb.StateMachineOperation
+
+	// Break up work into chunks and check context periodically
+	const chunkSize = 1000
+	for i := 0; i < len(l.operations); i += chunkSize {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+		}
+
+		end := i + chunkSize
+		if end > len(l.operations) {
+			end = len(l.operations)
+		}
+
+		for _, op := range l.operations[i:end] {
+			if op.OperationTime.AsTime().After(cutoff) {
+				retained = append(retained, op)
+			}
+		}
+	}
+	afterCount := len(retained)
+	l.logger.Info("Pruned operations",
+		tag.Name("oplog"),
+		tag.Value(fmt.Sprintf("pruned %d operations", beforeCount-afterCount)))
+
+	l.operations = retained
+	return nil
+}
+
+func (l *OpLog) Close() error {
+	l.Lock()
+	defer l.Unlock()
+
+	if l.closed {
+		return nil
+	}
+
+	l.logger.Info("Closing operation log",
+		tag.Name("oplog"),
+		tag.Value("close"))
+
+	close(l.stopCh)
+
+	// Wait for any in-progress prune operation to complete
+	l.wg.Wait()
+
+	l.closed = true
+	return nil
+}
+
+func (l *OpLog) pruneLoop() {
```

Comment:
These objects are automatically evicted from memory, there's no need to run pruning.